src/test/java/com/mongodb/hibernate/jdbc/MongoPreparedStatementTests.java

```diff
@@ -0,0 +1,335 @@
+/*
+ * Copyright 2024-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.jdbc;
+
+import static com.mongodb.hibernate.jdbc.MongoPreparedStatement.ERROR_MSG_PATTERN_PARAMETER_INDEX_INVALID;
+import static java.lang.String.format;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Answers.RETURNS_SMART_NULLS;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+
+import com.mongodb.client.ClientSession;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoDatabase;
+import java.sql.Blob;
+import java.sql.Clob;
+import java.sql.SQLException;
+import java.sql.Types;
+import java.util.Map;
+import java.util.stream.Stream;
+import org.bson.BsonDocument;
+import org.bson.Document;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.mockito.junit.jupiter.MockitoSettings;
+import org.mockito.quality.Strictness;
+
+@ExtendWith(MockitoExtension.class)
+@MockitoSettings(strictness = Strictness.WARN)
+class MongoPreparedStatementTests {
+
+    @Mock(answer = RETURNS_SMART_NULLS)
+    private MongoClient mongoClient;
+
+    @Mock(answer = RETURNS_SMART_NULLS)
+    private ClientSession clientSession;
+
+    @Mock(answer = RETURNS_SMART_NULLS)
+    private MongoConnection mongoConnection;
+
+    private MongoPreparedStatement createMongoPreparedStatement(String mql) {
+        return new MongoPreparedStatement(mongoClient, clientSession, mongoConnection, mql);
+    }
+
+    @Nested
+    class ParameterParsingTests {
```

Comment:
I don't think any of the tests in `ParameterParsingTests` are needed:

1. We should generally test the behavior of an object, not its internal state. However, the tests in `ParameterParsingTests` test the internal state of `MongoPreparedStatement` via the `@VisibleForTesting` method `getParameters`.
2. It's under exotic circumstances that we may resort to the `@VisibleForTesting` methods. And for `MongoPreparedStatement`, I don't see any reason for that.
3. If there are MQLs in `ParameterParsingTests` that test situations not covered by tests in `ParameterValueSettingTests`, then we should simply add tests to `ParameterValueSettingTests`.

Reply:
It is leftover when I tried to catch all the edge cases including:

1. same index was used more than once to set value
2. some index is not set at all

Later on I referred to other JDBC driver's implementation and found they invariably only test index range and never care about such edge cases. I simplified the logic accordingly but forgot to remove the associated unit tests.